encoder/libav_encoder.cpp

```diff
@@ -0,0 +1,548 @@
+
+	frame->buf[0] = av_buffer_alloc(sizeof(AVDRMFrameDescriptor));
+	frame->data[0] = frame->buf[0]->data;
+
+	AVDRMFrameDescriptor *desc = (AVDRMFrameDescriptor *)frame->data[0];
+	desc->nb_objects = 1;
+	desc->objects[0].fd = fd;
+	desc->objects[0].size = size;
+	desc->objects[0].format_modifier = DRM_FORMAT_MOD_INVALID;
+
+	desc->nb_layers = 1;
+	desc->layers[0].format = DRM_FORMAT_YUV420;
+	desc->layers[0].nb_planes = 3;
+	desc->layers[0].planes[0].object_index = 0;
+	desc->layers[0].planes[0].offset = 0;
+	desc->layers[0].planes[0].pitch = info.stride;
+	desc->layers[0].planes[1].object_index = 0;
+	desc->layers[0].planes[1].offset = info.stride * info.height;
+	desc->layers[0].planes[1].pitch = info.stride >> 1;
+	desc->layers[0].planes[2].object_index = 0;
+	desc->layers[0].planes[2].offset = info.stride * info.height * 5 / 4;
+	desc->layers[0].planes[2].pitch = info.stride >> 1;
+
+	std::scoped_lock<std::mutex> lock(video_mutex_);
+	frame_queue_.push(frame);
+	video_cv_.notify_all();
+}
+
+void LibAvEncoder::initOutput()
+{
+	int ret;
+
+	// Copy the global header from the video encode context once the first frame
+	// has been encoded.
+	avcodec_parameters_from_context(stream_[Video]->codecpar, codec_ctx_[Video]);
+
+	char err[64];
+	if (!(out_fmt_ctx_->flags & AVFMT_NOFILE))
+	{
+		ret = avio_open2(&out_fmt_ctx_->pb, options_->output.c_str(), AVIO_FLAG_WRITE, nullptr, nullptr);
+		if (ret < 0)
+		{
+			av_strerror(ret, err, sizeof(err));
+			throw std::runtime_error("libav: unable to open output mux for " + options_->output + ": " + err);
+		}
+	}
+
+	ret = avformat_write_header(out_fmt_ctx_, nullptr);
+	if (ret < 0)
+	{
+		av_strerror(ret, err, sizeof(err));
+		throw std::runtime_error("libav: unable write output mux header for " + options_->output + ": " + err);
+	}
+}
+
+void LibAvEncoder::deinitOutput()
+{
+	if (!out_fmt_ctx_)
+		return;
+
+	av_write_trailer(out_fmt_ctx_);
+
+	if (!(out_fmt_ctx_->flags & AVFMT_NOFILE))
+		avio_closep(&out_fmt_ctx_->pb);
+}
+
+void LibAvEncoder::encode(AVPacket *pkt, unsigned int stream_id)
+{
+	int ret = 0;
+
+	while (ret >= 0)
+	{
+		ret = avcodec_receive_packet(codec_ctx_[stream_id], pkt);
+
+		if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF)
+		{
+			av_packet_unref(pkt);
+			break;
+		}
+		else if (ret < 0)
+			throw std::runtime_error("libav error receiving packet: " + std::to_string(ret));
+
+		// Initialise the ouput mux on the first received video packet, as we may need
+		// to copy global header data from the encoder.
+		if (stream_id == Video && !init_output_)
+		{
+			initOutput();
+			init_output_ = true;
+		}
+
+		pkt->stream_index = stream_id;
+		pkt->pos = -1;
+		pkt->duration = 0;
+
+		// Rescale from the codec timebase to the stream timebase.
+		av_packet_rescale_ts(pkt, codec_ctx_[stream_id]->time_base, out_fmt_ctx_->streams[stream_id]->time_base);
+
+		std::scoped_lock<std::mutex> lock(output_mutex_);
+		// pkt is now blank (av_interleaved_write_frame() takes ownership of
```

Comment:
Am I right in thinking this mutex only ever gets grabbed once? That's feeling a bit weird to me...

Reply:
Ah, or do we have two threads calling this function?